lib/compress/zstd_double_fast.c

```diff
@@ -104,13 +102,15 @@ size_t ZSTD_compressBlock_doubleFast_generic(
 
     /* if a dictionary is attached, it must be within window range */
     if (dictMode == ZSTD_dictMatchState) {
-        assert(lowestValid + maxDistance >= endIndex);
+        assert(ms->window.dictLimit + (1U << cParams->windowLog) >= endIndex);
     }
 
     /* init */
     ip += (dictAndPrefixLength == 0);
     if (dictMode == ZSTD_noDict) {
-        U32 const maxRep = (U32)(ip - prefixLowest);
+        U32 const current = (U32)(ip - base);
+        U32 const windowLow = ZSTD_getLowestPrefixIndex(ms, current, cParams->windowLog);
+        U32 const maxRep = current - windowLow;
```

Comment:
@Cyan4973 This is the change that causes the compressed size difference.

Before we were invalidating repcodes for all indices that are invalid by the end of the block.

Now we are only invalidating repcodes that are beyond our maximum offset. When we don't have a dictionary this is clearly okay. When we have a dictionary this is still okay, because we know the dictionary is valid for the entire block.

This shows up as a big difference for small window sizes because each block is exactly 1 window size, which means that the repcodes are never valid.

Do you agree?

Reply:
> Before we were invalidating repcodes for all indices that are invalid by the end of the block.

OK, it took me a while, but I'm finally reading the code the same way.

In `maxRep = (U32)(ip - prefixLowest)`, the trick is that `prefixLowest` was determined relative to the _end_ of the block, to ensure than any match found with an index `>= prefixLowest` is necessarily valid anywhere in the entire block.

As a consequence, `windowLow` can be as close as the beginning of the block, when window size == block size, effectively invalidating all repcodes.
The new code `maxRep = current - windowLow` checks more thoroughly vs the actual window size (and of course valid range within the prefix). So it effectively increases the range for which repcodes are valid, up to a maximum of `windowSize`, which is correct.

Hence, with more opportunities to find repcode matches at the beginning of the block, it can translate into a few more bytes saved here and there.